### ISSUE TITLE
Why the keys are converted to string?

### ISSUE THREAD

**Issue by ecarreras** [honzajavorek/redis-collections#25]
Python dict:
```python
d = {}
d[1] = 'test'
{1: 'test'}
```

With RedisDict:
```python
d = Dict(redis=r)
d[1] = 'test'
{'1': 'test'}
```

Is the same issue that https://github.com/disqus/durabledict/issues/4

**Comment by bbayles** [comment]
Pickling the keys was the first thing I thought of also, but this introduces a problem* - keys that are equal and have the same hash don't necessarily pickle to the same byte string.

For example, you can't have a `dict` with `1.0` and `1` as the same key, but they pickle to different values.
```python
>>> import pickle
>>> 1.0 == 1
0: True
>>> hash(1.0) == hash(1)
1: True
>>> pickle.dumps(1.0) == pickle.dumps(1)
2: False
>>> len({1.0: 'x', 1: 'y'})
3: 1
```

To fix this issue I think a bigger change is called for. Instead of storing keys and pickled values, I propose storing hashes and pickled dicts. For example:

```python
getitem(redis, collection, key)
    key_hash = hash(key)
    D = redis.hget(collection, key_hash)
    D = {} if D is None else loads(D)
    return D[key]

setitem(redis, collection, key):
    key_hash = hash(key)
    D = redis.hget(collection, key_hash)
    D = {} if D is None else loads(D)
    D[key] = value
    redis.hset(collection, key_hash, dumps(D))
```

This restores correct behavior - `1.0` and `1` are the same key.

---

\* This problem exists now, even without pickling keys.

```python
>>> from redis_collections import Dict
... redis_dict = Dict()
... python_dict = {}
... 
... for D in (redis_dict, python_dict):
...     D[1] = 'x'
...     D[1.0] = 'y'
... 
... print(redis_dict.items(), python_dict.items())
[('1', 'x'), ('1.0', 'y')] [(1, 'y')]
```

**Comment by bbayles** [comment]
Closed via #45, I think!
